app/scripts/migrations/077-supplements/077-supplements.md

```diff
@@ -0,0 +1,98 @@
+data could overwrite the existing `networkConfigurations` data, and the user could lose `networkConfigurations` data that had been written to their state since migration 82 had first failed to run.
+
+To fix this, the migration 82 supplement deletes `frequentRpcListDetail` if the `networkConfigurations` object exists. Users in such a scenario will have network data in `networkConfigurations` that
+they have been using, while the `frequentRpcListDetail` data would not have been seen for some time. So the best thing to do for them is delete their old data and preserve the data they have most recently
+used.
+
+**Migration 84**
+
+Replaces the `NetworkController.network` property with a `networkId` and `networkStatus` property. If this migration had not run, the NetworkController would have a `network` property and
+`networkId` and `networkStatus` properties. If migration 84 later ran on this state, the old (and forgotten) `network` property could cause the `networkId` and `networkStatus` to be overwritten,
+affecting the state the user's experience was currently depending on.
+
+The fix in the migration 84 supplement is to delete the `NetworkController.network` property if the `NetworkId` property already exists.
+
+**Migration 86**
+
+Renamed the `NetworkController.provider` property to `providerConfig`. If this migration had not run, the NetworkController would have a `provider` property and
+a `providerConfig` property. If migration 86 later ran on this state, the old (and forgotten) `provider` property could cause the `providerConfig` property to be overwritten,
+affecting the state the user's experience was currently depending on.
+
+The fix in the migration 86 supplement is to delete the `NetworkController.provider` property if the `providerConfig` property already exists.
+
+**Migration 88**
+
+Attempted to change the keys of multiple parts of state related to tokens. In particular, `NftController.allNftContracts`, `NftController.allNfts`, `TokenListController.tokensChainsCache`, `TokensController.allTokens`, `TokensController.allIgnoredTokens` and `TokensController.allDetectedTokens`. All of these objects were keyed by chainId in decimal number form. The migration's
+purpose was to change those decimal chain ID keys to hexadecimal. If migration 77 failed, and then the user added or modified tokens, they could have duplicates within these parts of state:
+some with decimal keys and others with an equivalent hexadecimal key. If the data pointed to by those keys was modified at all, and the migration 88 was later run, the most recent data (under
+the hexadecimal key) could be overwritten by the old data under the decimal key.
+
+The migration 88 supplement fixes this by deleting the properties with decimal keys if an equivalent hexadecimal key exists.

```

Comment:
We could consider deleting all decimal-keyed entries instead. Currently we're restoring the asset data that had previously been lost, which might be greatly appreciated by users but might also be confusing and undesirable.

Reply:
I actually agree with deletion, for the same reasons I presented in response to lego's comment above. So I made that change in e67aabb70b

Reply:
Another alternative would be to take the `lostidentities` approach, and quarantine this old invalid state for now. Then we could delete it later, or merge it instead if we decide that's preferable. I'm OK with deleting it though.

Reply:
> I actually agree with deletion, for the same reasons I presented in response to lego's comment above. So I made that change in https://github.com/MetaMask/metamask-extension/commit/e67aabb70bb118110451229aa69f136f902729b2

The unit tests failed due to a change in that commit. The failure captures the case where a user had not yet updated to a version where hexadecimal entries were being added to state. In that case, we might have deleted all of a users tokens on update to v10.34.2

So I think what we need to do is delete all decimal-keyed entries if there is a single hexadecimal-keyed entry. But if not, assume that the migration is running for the first time for the user, and don't delete in that case.

Reply:
> So I think what we need to do is delete all decimal-keyed entries if there is a single hexadecimal-keyed entry. But if not, assume that the migration is running for the first time for the user, and don't delete in that case.

I applied that change in f8c5acbf1e

Reply:
> Perhaps we should have one central check for the "future state" case, so that we're consistent in how we handle this across each controller and supplemental migration.

@Gudahtt Did you have any ideas for what this check should be?

Reply:
Perhaps if any of the future state conditions are true, then apply the deletes accross all supplements?

So if `networkConfigurationIsPresent || hexChainIdKeyedTokenCollectionsArePresent || networkIdIsPresent || providerConfigIsPresent` then we apply it?

Reply:
Yes exactly, that's what I was thinking

Reply:
> Hmm. Perhaps we should have one central check for the "future state" case, so that we're consistent in how we handle this across each controller and supplemental migration.

> We wouldn't want to arbitrarily restore some state but not others, just because users haven't used a given state collection since the problem started.

Actually, I am not sure that we can do this in a good way. Suppose a user in the 077 error state has upgraded to a version that included migrations 82, 84 and 86, but was pre migration 88. They've got nfts and tokens that they are using. In state these would be keyed by decimal chain IDs. Then they later update to v10.34.2: we would delete all their tokens because of the network state, even though we did not need to.

Reply:
Hmm. Good point. The approach you've taken here is probably best then.